uniffi_bindgen/src/scaffolding/templates/ErrorTemplate.rs

```diff
@@ -1,20 +1,48 @@
 {#
-    // For each error declared in the UDL, using the [Error] attribute, we assume the caller has provided a corresponding
-    // rust Error enum with the same name. We provide the traits for sending it across the FFI, which will fail to
-    // compile if the provided enum has a different shape to the one declared in the UDL. 
-    // Here we define the neccessary converstion to allow the error to propegate through the FFI as an error.
+// For each enum declared in the UDL, we assume the caller has provided a corresponding
+// rust `enum`. We provide the traits for sending it across the FFI, which will fail to
+// compile if the provided struct has a different shape to the one declared in the UDL.
 #}
 #[doc(hidden)]
-impl From<{{e.name()}}> for uniffi::deps::ffi_support::ExternError {
-    fn from(err: {{e.name()}}) -> uniffi::deps::ffi_support::ExternError {
-        // Errno just differentiate between the errors.
-        // They are in-order, i.e the first variant of the enum has code 1
-        // As we add support for generic errors (e.g panics) 
-        // we might find that we need to reserve some codes.
-        match err {
-            {%- for value in e.values() %}
-            {{ e.name()}}::{{value}}{..} => uniffi::deps::ffi_support::ExternError::new_error(uniffi::deps::ffi_support::ErrorCode::new({{ loop.index }}), err.to_string()),
+impl uniffi::RustBufferViaFfi for {{ e.name() }} {
+    fn write(&self, buf: &mut Vec<u8>) {
+        use uniffi::deps::bytes::BufMut;
+        match self {
+            {%- for variant in e.variants() %}
+            {{ e.name() }}::{{ variant.name() }}{% if variant.has_fields() %} { {% for field in variant.fields() %}{{ field.name() }}, {%- endfor %} }{% else %}{..}{% endif %} => {
+                buf.put_i32({{ loop.index }});
+                {% for field in variant.fields() -%}
+                <{{ field.type_()|type_rs }} as uniffi::ViaFfi>::write({{ field.name() }}, buf);
+                {%- endfor %}
+            },
             {%- endfor %}
-        }
+        };
     }
+
+    {% if e.is_flat() %}
+    // If a variant doesn't have fields defined in the UDL, it's currently still possible that
+    // the Rust enum has fields and they're just not listed.  Let's just punt on implemented
+    // try_read() in that case, which is no issue since passing back Errors into the rust code
+    // isn't supported.
+    fn try_read(_buf: &mut &[u8]) -> uniffi::deps::anyhow::Result<Self> {
+        panic!("try_read not supported for fieldless errors");
```

Comment:
This was kind of unfortunate, but I think it shouldn't cause issues unless someone wants to pass an error from the consumer code into the rust code.  Going forward, maybe we could deprecate error enums with associated data that don't define that data in the UDL file.

Reply:
Agreed, this seems safe to punt on, and deprecating it seems like a good path forward.

Reply:
Your comments reminded me that we only support Error enums with named fields, which makes me hesitant to deprecate this and force consumers to change how they structure their errors.  I'm thinking we should just not worry about this corner case for now.

Reply:
Ah yes, fair enough. I agree it's seems very unlikely to cause us problems in practice, happy to just punt on it for now.